generator/internal/discovery_type_vertex_test.cc

```diff
@@ -763,6 +764,436 @@ message ReservedHighest {
+      StatusIs(
+          StatusCode::kInvalidArgument,
+          HasSubstr(
+              "Message: generator.test.FieldsOnly has field: field6 "
+              "whose type has changed from: map<string, "
+              "generator.imported.Bar> to: map<string, generator.test.Bar>")));
+}
+
+TEST_F(DiscoveryTypeVertexDescriptorTest,
+       JsonToProtobufMessageWithReservedFieldNumberLessThanMax) {
+  auto constexpr kPreviousProto = R"""(
+syntax = "proto3";
+package test.package;
+// Description of the message.
+message TestSchema {
+  reserved 1;
+  // Description for existingField.
+  optional string existing_field = 2;
+}
+)""";
+
+  auto constexpr kSchemaJson = R"""(
+{
+  "description": "Description of the message.",
+  "type": "object",
+  "id": "TestSchema",
+  "properties": {
+    "newField": {
+      "$ref": "Foo",
+      "description": "Description for newField."
+    },
+    "existingField": {
+      "type": "string",
+      "description": "Description for existingField."
+    }
+  }
+}
+)""";
+
+  auto constexpr kExpectedProto = R"""(// Description of the message.
+message TestSchema {
+  reserved 1;
+  // Description for existingField.
+  optional string existing_field = 2;
+
+  // Description for newField.
+  optional Foo new_field = 3;
+}
+)""";
+
+  ASSERT_TRUE(AddProtoFile("test.proto", kPreviousProto));
+  auto json = nlohmann::json::parse(kSchemaJson, nullptr, false);
+  ASSERT_TRUE(json.is_object());
+  DiscoveryTypeVertex t("TestSchema", "test.package", json, &pool());
+  std::map<std::string, DiscoveryTypeVertex> types;
+  types.emplace("Foo", DiscoveryTypeVertex{"Foo", "test.package", {}, &pool()});
+  auto result = t.JsonToProtobufMessage(types, "test.package");
+  ASSERT_STATUS_OK(result);
+  EXPECT_THAT(*result, Eq(kExpectedProto));
+}
+
+TEST_F(DiscoveryTypeVertexDescriptorTest,
+       JsonToProtobufMessageWithReservedFieldNumberGreaterThanMax) {
+  auto constexpr kPreviousProto = R"""(
+syntax = "proto3";
+package test.package;
+// Description of the message.
+message TestSchema {
+  reserved 3;
+  // Description for existingField.
+  optional string existing_field = 1;
+
+  // Description for otherExistingField.
+  map<string, string> other_existing_field = 2;
+}
+)""";
+
+  auto constexpr kSchemaJson = R"""(
+{
+  "description": "Description of the message.",
+  "type": "object",
+  "id": "TestSchema",
+  "properties": {
+    "newMapField": {
+      "additionalProperties": {
+        "type": "string"
+      },
+      "type": "object",
+      "description": "Description for newMapField."
+    },
+    "existingField": {
+      "type": "string",
+      "description": "Description for existingField."
+    },
+    "otherExistingField": {
+      "additionalProperties": {
+        "type": "string"
+      },
+      "type": "object",
+      "description": "Description for otherExistingField."
+    }
+  }
+}
+)""";
+
+  auto constexpr kExpectedProto = R"""(// Description of the message.
+message TestSchema {
```

Comment:
These expectations are fairly strict.  I assume the ordering of the messages is guaranteed?  Alphabetical perhaps?  Should they be generated ordered by message number?